### PR TITLE
DEP: Deprecate `data_type.dtype` if attribute is not already a dtype

### PR DESCRIPTION
**Pull request by seberg**
Followup from gh-13003, ~marking as draft because that one needs to be merged (and then this one rebased) before it can go in.~

This could hit the mailing list, but I think the deprecation here should be pretty uncontroversial, since it only enforces that `something.dtype` returns a dtype in statements such as `np.dtype(something)`. It does not actually force the user to write the `.dtype` yet.

### CODE REVIEW

**Comment by BvB93** [comment]
A question: is it the plan to (eventually) move forward with this PR or is that still undecided?

**Comment by seberg** [comment]
Not sure why it stalled, unless anyone knows of a  user who expects `obj.dtype.dtype` to be called, and even if, this is just a deprecation.

File: numpy/core/src/multiarray/descriptor.c [hunk 1]
```diff
@@ -122,6 +127,16 @@ _try_convert_from_dtype_attr(PyObject *obj)
         goto fail;
     }
 
+    if (DEPRECATE("in the future the `.dtype` attribute of a given data"
+                  "type object must be a valid dtype instance. "
+                  "`data_type.dtype` may need to be coerced using "
+                  "`np.dtype(data_type.dtype)`. This will raise an error "
+                  "in NumPy 1.19.") < 0) {
+        /* NumPy 1.19, 2020-02-07 */
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c:135]:
```suggestion
                  "`np.dtype(data_type.dtype)`.") < 0) {
        /* NumPy 1.20, 2020-10-19 */
```

**Comment by BvB93** [comment]
IMO it seems like a reasonable change and it makes typing a bit easier, so +1.

**Comment by mattip** [comment]
hmm. The circelCI merge-before-build command seems to be causing problems. Maybe we should allow it to fail silently.

**Comment by mattip** [comment]
The deprecation warning is leaking out into tests

**Comment by seberg** [comment]
@BvB93 its a new typing test that is failing (and not running locally with the slow flag).  Should the test be modified or catch the warning?

**Comment by BvB93** [comment]
> Should the test be modified or catch the warning?

Ah yes, the `dtype` value should be changed from `float` to something along the lines of `np.dtype(float)`:

https://github.com/numpy/numpy/blob/136185cee387e0e3638abf38b4faf9b0f0cfadd7/numpy/typing/tests/data/pass/dtype.py#L31-L32

**Comment by seberg** [comment]
Thanks! Just wasn't sure which direction to fix it best.

**Comment by seberg** [comment]
@mattip I think the circleci is probably only because I had it set up on my branch.  Disabled it now, it is not something I use regularly anyway.

File: numpy/f2py/tests/test_array_from_pyobj.py [hunk 2]
```diff
@@ -3,9 +3,8 @@
 import copy
 import pytest
 
-from numpy import (
-    array, alltrue, ndarray, zeros, dtype, intp, clongdouble
-    )
+import numpy as np
+
```
Comment (mattip) [inline numpy/f2py/tests/test_array_from_pyobj.py:7]: Is this change required to avoid the deprecation warning? Does that mean all user code must now change?

Comment (seberg) [reply numpy/f2py/tests/test_array_from_pyobj.py:7]: No, don't remember why I did it here, but it is just maintanence, there must have been an issue with dtype being overshadowed somewhere.

**Comment by mattip** [comment]
Could you move the deprecation test to the proper file? Otherwise this looks good to me.

**Comment by seberg** [comment]
Done, we could consider deprecating even more, but I guess this might be a good (and easy) first step.

Yeah, I think I first wrote that at a time when I was considering to stop relying on that file so much (The original reason, and most of the magic in that file, was to work around python issues with warnings, which were eventually fixed).

**Comment by mattip** [comment]
Thanks @seberg

**Comment by charris** [comment]
@seberg I notice that this created `changelog/13578.deprecation.rst` in the root directory. What is that about? Should it be in the release notes?

**Comment by seberg** [comment]
Oh wow, that is a while ago. Yeah wrong place, should have been part of the release notes...

**Comment by charris** [comment]
OK, I'll make a PR to get rid of it in main and take care of it in the 1.21.x branch.
